File: dsa/two-pointers/42. Trapping Rain Water.cpp
/*
LC 42 — Trapping Rain Water
Pattern: Two Pointers
Difficulty: Hard
Approach: Use two pointers at both ends. Maintain leftMax and rightMax.
          The amount of water at any index is determined by the shorter of the 
          two boundaries. Move the pointer pointing to the smaller maximum height 
          and calculate trapped water for that spot.
Time Complexity: O(n)
Space Complexity: O(1)
*/
#include<bits/stdc++.h>
#include <iostream>
using namespace std;

class Solution {
public:
    int trap(vector<int>& height) {
        if (height.empty()) return 0;
        
        int l = 0, r = height.size() - 1;
        int leftMax = height[l], rightMax = height[r];
        int ans = 0;

        while (l < r) {
            if (leftMax < rightMax) {
                l++;
                leftMax = max(leftMax, height[l]);
                ans += leftMax - height[l];
            } else {
                r--;
                rightMax = max(rightMax, height[r]);
                ans += rightMax - height[r];
            }
        }
        return ans;
    }
};

File: dsa/two-pointers/42. Trapping Rain Water.py
"""
LC 42 — Trapping Rain Water
Pattern: Two Pointers
Difficulty: Hard
Approach: Use two pointers at both ends. Maintain left_max and right_max.
          Water trapped is limited by the shorter boundary. 
          If left_max is smaller, we know the bottleneck for the current left index
          is left_max, so we calculate water and move left. Otherwise, move right.
Time Complexity: O(n)
Space Complexity: O(1)
"""
from typing import List

class Solution:
    def trap(self, height: List[int]) -> int:
        if not height:
            return 0
            
        l, r = 0, len(height) - 1
        left_max, right_max = height[l], height[r]
        ans = 0
        
        while l < r:
            if left_max < right_max:
                l += 1
                left_max = max(left_max, height[l])
                ans += left_max - height[l]
            else:
                r -= 1
                right_max = max(right_max, height[r])
                ans += right_max - height[r]
                
        return ans